### ISSUE TITLE
Features to help make a 3D only user-interface easier . .

### ISSUE THREAD

**Issue by charlesmlamb** [godotengine/godot-proposals#3418]
### Describe the project you are working on

A barbarian game, 2.5D . . . .

### Describe the problem or limitation you are having in your project

I want to have an UI, with real 3D elements, that is not currently possible, using the Control nodes, the only way to actually do that, is to use a sort of 3D UI, by placing planes, with textures, in front of the screen . . . The problem I run into, is it is difficult to align the planes, to the FOV, also any changes later, to this setting, requires re-aligning the planes . . .

I was sort-of thinking, might one say make a node, one could attach to a camera, that could add some planes, for use in a 3D UI, since 2D is a bit, well atm. lacking, over having true 3D, in the interface . . .

Here's a demo . . .

![Skærmbillede (2365)](https://user-images.githubusercontent.com/83387915/137026612-aa064a29-68ec-4b96-b029-e5db51d1f71e.png)

Was thinking, since one sometimes want a 3D only UI, not using control nodes, also to get stuff like 3D rendered objects in the UI, anyway one can make some cool UI elements this way, the nice thing is, if the planes sort-of snapped to the FOV, one would have, pixel-perfect UI, and could design a normal 2D UI, and add that to a 3D UI, doing it by hand, tends to lead to crazy calculations, so on . . . And, if one changes FOV, it's also new math . . .

The suggestion is, make some sort of 3D UI editor, perhaps some new nodes, that one can make under, a Camera, and one can set, for instance if total resolution is 1280x720, that a plane has an offset, of 200 pixels, and is 400x540 pixels, so one can make a layered 3D UI, more easily, so on . . .

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Well, I need to make a 3D UI, 100 %, to get 3D elements in the UI, like some modern 3D games . . But, also need the ability, to make the screen a bit dim, using a plane behind the 3D UI object, like, an item, say a ring, something, etc . . . Then, increase opacity, to get a dim effect, while the 3D object, a pick - up, retains normal colors . . . This is much easier in 3D, however atm. it's manual work, to calculate the planes, so on, ideally one wants pixel perfect, so the planes snap to FOV, to get a ' best - case ' 3D only UI, maybe there's a way to make this remotely possible, some new nodes, Control3D nodes, not sure . . . .

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Well, add a setting, to make a 3D UI, since that allows for some nice UI, and it'd be pixel perfect, if the planes were calculated correctly, imo . . . Right now, that's quite difficult, say if one wants a plane, that is offset 300, 200 pixels, and it size 440x580 pixels, it'd be nice, to have some help, or some simple editor, to do that . . .

Here's an example, from a PS3 game, where there are real 3D elements, in the UI, something that's not possible in the Control nodes, was thinking it might be possible, to help 3D artists, that want to make 3D only games, to make a nice UI, this way . . . Note also, with a bit of help, it'd be possible to make the planes, for UI, rather small, like 16x9 cm, in the game-world, so they'd not clip, or be really visible, in the editor, bec. they'd be small, compared to world, so on . . . .

![2021-10-12 2254](https://user-images.githubusercontent.com/83387915/137028061-7a154774-c3d5-4feb-9f65-7b4091f328ed.jpg)

So, some way to add planes, set distance from camera, perhaps add stuff like meshinstance points, set scale, position, other planes, to create a ' pixel perfect ', or close 3D UI editor, or so . . .

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It'd possible, for stuff like a plane, that fills the FOV 100 %, but for say a plane, with an offset, in pixels, the math gets really complex, so it's not easy to do, a simple 3D UI, or perhaps Control3D editor, would make sense for some 3D titles, maybe, most of them, at least for me, it's about getting real, not 3D-to-2D UI elements, bec. that looks rather nice, also, found in many games, going back 10 years, etc . . . .

### Is there a reason why this should be core and not an add-on in the asset library?

Well, the current 2D UI nodes, don't meet all use - cases, making it oneself takes time, and it'd allow better 3D UI, for 3D games, and if one could add it to a 2D game, it'd allow a 3D camera, to place 3D objects in front of a 2D game, maintaining pixel ratio, or so . . :OO

**Comment by charlesmlamb** [comment]
No, the coin is 3D, it rotates, in the game . . One can make a viewporttexture, and render it, from 3D to 2D, in Godot, but to get true 3D elements, on top of other UI elements, one can't use control nodes . . .

Sort-of asking, for an easier way, to set it up, and get 1-to-1 pixel accurate renders, etc . . . !

**Comment by SIsilicon** [comment]
> No, the coin is 3D, it rotates, in the game . . One can make a viewporttexture, and render it, from 3D to 2D, in Godot, but to get true 3D elements, on top of other UI elements, one can't use control nodes . . .

You can render 3D over UI, and I just explained to you how. The coin I mentioned is still a 3D object, so you can still rotate it like any 3D object.
